lib/Mojolicious/Guides/Testing.pod

```diff
@@ -0,0 +1,754 @@
+
+  use Test::Mojo;
+  use Test::More;
+
+  # Start a Mojolicious app named "Celestial"
+  my $t = Test::Mojo->new('Celestial');
+
+  # Post a JSON document
+  $t->post_ok('/notifications',
+              json => { event   => "full moon",
+                        via     => "sms",
+                        channel => '+15555555555' })
+    ->status_is(201)
+    ->json_is('/message' => "notification created");
+
+  # Perform GET requests and look at the responses
+  $t->get_ok('/sunrise')
+    ->status_is(200)
+    ->content_like(qr/ am$/);
+
+  $t->get_ok('/sunset')
+    ->status_is(200)
+    ->content_like(qr/ pm$/);
+
+  # Post a URL-encoded form
+  $t->post_ok('/insurance',
+              form => { name   => 'Jimmy',
+                        amount => '3.000.000',
+                        currency => '€' })
+    ->status_is(200);
+
+  # Use Test::More's like() to check the response
+  like $t->tx->res->dom->at('div#thanks')->text, qr(thank you), "thanks";
+
+  done_testing();
+
+In the rest of this document we'll explore these concepts and others related
+to L<Test::Mojo>.
+
+=head1 CONCEPTS
+
+Essentials every L<Mojolicious> developer should know.
+
+=head2 L<Test::Mojo> At a Glance
+
+The L<Test::More> module bundled with Perl includes several primitive test
+assertions, such as C<ok>, C<is>, C<isnt>, C<like>, C<unlike>, C<cmp_ok>,
+etc. An assertion "passes" if its expression returns a true value. The
+assertion method prints "ok" or "not ok" if an assertion passes or fails
+(respectively).
+
+L<Test::Mojo> supplies additional test assertions organized around the web
+application request/response transaction (transport, response headers,
+response bodies, etc.), and WebSocket communications.
+
+One interesting thing of note: the return value of L<Test::Mojo> object
+assertions is always the test object itself, allowing us to "chain" test
+assertion methods. So rather than grouping related test statements like this:
+
+  $t->get_ok('/frogs');
+  $t->status_is(200);
+  $t->content_like(qr/bullfrog/);
+  $t->content_like(qr/hypnotoad/);
+
+method chaining allows us to connect test assertions that belong together:
+
+  $t->get_ok('/frogs')
+    ->status_is(200)
+    ->content_like(qr/bullfrog/)
+    ->content_like(qr/hypnotoad/);
+
+This makes for a much more I<concise> and I<coherent> testing experience:
+concise because we are not repeating the invocant for each test, and coherent
+because assertions that belong to the same request are syntactically bound in
+the same method chain.
+
+Occsionally it makes sense to break up a test to perform more complex
+assertions on a response. L<Test::Mojo> exposes the entire transaction object
+so you can get all the data you need from a response:
+
+  $t->put_ok('/bees', json => { type => 'worker', name => 'Karl' })
+    ->status_is(202)
+    ->json_has('/id');
+
+  # Pull out the id from the response
+  my $newbee = $t->tx->res->json('/id');
+
+  # Make a new request with data from the previous response
```

Comment:
This is great to share that you can do this.  My only comment is that I've been told (my understanding may be way wrong) that tests shouldn't be built this way.  I don't fully understand why.  @jberger gave me the advice.  If possible, I would love to have better explanations to *how* tests should be made (links to materials to read, etc).  I know it's not exactly Mojolicious-specific, perhaps a reference to basic Testing advice would fit in the opening doc that points us to Learning Perl in 2.5 hours.
I don't understand good testing techniques, I just read the Test::Mojo docs and did what they allowed, they allowed me to do what's suggested here on L109, but @jberger suggested this wasn't a good way to design tests.  Again, I may be misunderstanding.  Any help in this regard would be wonderful, it would help users of Mojolicious to write really high quality apps that Mojolicious can be proud to be behind.  It's not good for Mojolicious to have bad apps written (just look at the stigma associated with Perl from early on!)